### DISCUSSION TITLE
Icat implementation

### DISCUSSION THREAD

**Discussion by noahfraiture** [kovidgoyal/kitty#7822] [Q&A]
Hello, I want to make a TUI in Go that display a generated image at some point. I searched among existing implementation but none of them works with gif so I though I could fork one and make an implementation for gif. I made a fork for rasterm (I changed public api so I created a new repo [igo](https://github.com/noahfraiture/igo)). But I can't make it work. Here's my current code : 
```
// Serialize GIF image into Kitty terminal in-band format.
func KittyWriteGIF(out io.Writer, gifReader io.Reader, opts KittyImgOpts) error {
	frames, delays := DecodeGIF(gifReader)
	if frames == nil {
		return errors.New("failed to decode GIF")
	}

	opts.SrcHeight = uint32(frames[0].Bounds().Dx())
	opts.SrcWidth = uint32(frames[0].Bounds().Dy())

	// Use the first frame as the base image
	if err := KittyWriteImage(out, frames[0], opts); err != nil {
		return err
	}

	// Write each frame with appropriate frame numbers
	for i, frame := range frames[1:] {
		if err := KittyWriteFrame(out, frame, opts, delays[i]); err != nil {
			return err
		}
	}

	return nil
}

func DecodeGIF(r io.Reader) ([]image.Image, []int) {
	img, err := gif.DecodeAll(r)
	if err != nil {
		fmt.Println("Error decoding GIF:", err)
		return nil, nil
	}

	frames := make([]image.Image, len(img.Image))
	delays := make([]int, len(img.Delay))
	for i, frame := range img.Image {
		frames[i] = frame
		delays[i] = img.Delay[i]
	}

	return frames, delays
}

func KittyWriteFrame(out io.Writer, frame image.Image, opts KittyImgOpts, delay int) error {
	// Ensure width and height are properly set
	if frame.Bounds().Dx() == 0 || frame.Bounds().Dy() == 0 {
		return errors.New("frame width or height is zero")
	}

	headerOpts := []string{
		"a=f",
		"t=d",
		"m=1",
		fmt.Sprintf("s=%d", opts.SrcWidth),
		fmt.Sprintf("v=%d", opts.SrcHeight),
		fmt.Sprintf("i=%d", opts.ImageId),
		fmt.Sprintf("z=%d", delay*10),
	}
	fmt.Println(headerOpts)

	_, err := fmt.Fprint(out, opts.ToHeader(headerOpts...))
	if err != nil {
		return err
	}

	buf := new(bytes.Buffer)
	if err := png.Encode(buf, frame); err != nil {
		return err
	}

	chunk := kittyChunk{
		chunkSize: 4096,
		writer:    out,
	}

	enc64 := base64.NewEncoder(base64.StdEncoding, &chunk)
	_, err = io.Copy(enc64, buf)
	return errors.Join(
		err,
		enc64.Close(),
		chunk.Close(),
	)
}
```
But I keep having the error ```.Gi=1159361787;OK.Gi=1159361787,r=2;ENODATA:Insufficient image data: 0 < 27604.Gi=1159361787,r=2;ENODATA:Insufficient image data: 0 < 27604.Gi=1159361787,r=2;ENODATA:Insufficient image data: 0 < 27604```  And I don't know where it comes from. Is there anyone that know how to solve this issue ?

**Comment by kovidgoyal** [comment]
ENODATA means the amount of pixel data you supplied is not enough for the specfied image/frame width,height and format.

**Comment by noahfraiture** [comment]
Thank you for that clarification.
Do you know where the problem could come from ? The write method used works for png or else and is the following ```func (c *kittyChunk) Write(buf []byte) (int, error) {
	bytesLeft := len(buf)
	written := 0

	for bytesLeft > 0 {

		toWrite := c.chunkSize
		if toWrite > bytesLeft {
			toWrite = bytesLeft
		}

		// prefix
		if _, err := fmt.Fprint(c.writer, KITTY_IMG_HDR, "m=1;"); err != nil {
			return written, err
		}

		// data
		n, err := c.writer.Write(buf[:toWrite])
		written += n
		if err != nil {
			return written, err
		}

		// suffix
		if _, err := fmt.Fprint(c.writer, KITTY_IMG_FTR); err != nil {
			return written, err
		}

		buf = buf[toWrite:]
		bytesLeft -= toWrite
	}

	return written, nil
}```

We enter in the function the `buf` isn't empty. Is there something obvious I am missing ?

**Comment by noahfraiture** [comment]
I tried to decompose the gif into multiple images but I still have the error `ENODATA:Insufficient image data: 170446 < 553776`.
There's something I don't understand about the gif format and the way kitty it. What is the problem ?
